### PR TITLE
View annotations: final core fixes

### PR DESCRIPTION
**Pull request by kiryldz**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog></changelog>`.

### Summary of changes

- Move `ViewAnnotationManagerImpl` to `com.mapbox.maps`. Some public records are now also internal.
- Add fix for view jumping.

| Before | After | 
| ----- | -----|
| <video src='https://user-images.githubusercontent.com/15800566/141281400-0e418995-90f9-4768-befc-2dde138cce8c.mp4' width=250/>  | <video src='https://user-images.githubusercontent.com/15800566/141281102-6a4e0624-1392-4c01-a74b-543715b2bb49.mp4  ' width=250/>|

 - Use translationX / translationY instead of setting margin of the view group.

### CODE REVIEW

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotation.kt [hunk 1]
```diff
@@ -16,6 +16,10 @@ internal data class ViewAnnotation(
    * if user did not specify [ViewAnnotationOptions.visible]  explicitly  during add / update operation
    */
   var handleVisibilityAutomatically: Boolean,
+  /**
+   * Cached value if view was visible before OnGlobalLayoutListener was triggered.
+   */
+  var wasVisible: Boolean,
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotation.kt:22]: what is the default value?
can user change this variable?

Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotation.kt:22]: this class is internal, we control it ourselves
see no sense in setting default value

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 2]
```diff
@@ -1,18 +1,19 @@
-package com.mapbox.maps.viewannotation
+package com.mapbox.maps
```
Comment (pengdev) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:1]: any reason to move it to another package?

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 3]
```diff
@@ -130,11 +132,21 @@ internal class ViewAnnotationManagerImpl(
     val viewAnnotation = ViewAnnotation(
       view = inflatedView,
       handleVisibilityAutomatically = options.visible == null,
+      wasVisible = true,
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:135]:
```suggestion
      wasVisible = inflatedView.visibility == View.VISIBLE,
```

**Review by alexander-kulikovskii — APPROVED** [review]

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt (hunk 2 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:1]: Kotlin could not see interface located in sub-package. As all internal things are generated in `com.mapbox.maps` - we need to move our class implementing this interface in same package as well.

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt (hunk 3 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt]: great catch, applied it 💎

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt (hunk 2 above)
Comment (pengdev) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:1]: Thanks for clarification.

**Review by pengdev — APPROVED** [review]

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 4]
```diff
@@ -130,11 +132,21 @@ internal class ViewAnnotationManagerImpl(
     val viewAnnotation = ViewAnnotation(
       view = inflatedView,
       handleVisibilityAutomatically = options.visible == null,
+      wasVisible = inflatedView.visibility == View.VISIBLE,
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:135]: IMO just `isVisible` would work better, it becomes `was` only in future in onGlobalLayout, I suppose

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 5]
```diff
@@ -130,11 +132,21 @@ internal class ViewAnnotationManagerImpl(
...
       viewLayoutParams = inflatedViewLayout,
     )
     val globalLayoutListener = ViewTreeObserver.OnGlobalLayoutListener {
       if (viewAnnotation.handleVisibilityAutomatically) {
         val isVisibleNow = inflatedView.visibility == View.VISIBLE
+        if (isVisibleNow == viewAnnotation.wasVisible) {
+          return@OnGlobalLayoutListener
+        }
+        viewAnnotation.wasVisible = isVisibleNow
+        // hide view below map surface and pull it back when new position from core will arrive
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:145]: Seems the hack deserve better explanation

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 6]
```diff
@@ -130,11 +132,21 @@ internal class ViewAnnotationManagerImpl(
...
+        if (isVisibleNow) {
+          hiddenViewMap[inflatedView] = inflatedView.translationZ
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:147]: Can't translationZ just be set to the view right now? Why is the map needed?

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt (hunk 4 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt]: done

**Pull request merged**